File: cpp/data_structures/linear/array/dynamic_array.cpp
// 手写动态数组（封装 vector + 显式扩容演示）
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class DynamicArray {
    vector<int> data_;
    size_t cap_;

    void ensure_extra(size_t extra) {
        if (data_.size() + extra > cap_) {
            cap_ = max(cap_ * 2, data_.size() + extra);
            data_.reserve(cap_);
        }
    }

public:
    explicit DynamicArray(size_t initial_capacity = 4) : cap_(max<size_t>(1, initial_capacity)) {
        data_.reserve(cap_);
    }

    size_t size() const { return data_.size(); }
    size_t capacity() const { return cap_; }

    int at(size_t index) const {
        if (index >= data_.size()) throw out_of_range("at");
        return data_[index];
    }

    void push_back(int v) {
        ensure_extra(1);
        data_.push_back(v);
    }

    int pop_back() {
        if (data_.empty()) throw underflow_error("pop_back");
        int v = data_.back();
        data_.pop_back();
        return v;
    }

    void insert(size_t index, int v) {
        if (index > data_.size()) throw out_of_range("insert");
        ensure_extra(1);
        data_.insert(data_.begin() + (ptrdiff_t)index, v);
    }

    void erase(size_t index) {
        if (index >= data_.size()) throw out_of_range("erase");
        data_.erase(data_.begin() + (ptrdiff_t)index);
    }
};

int main() {
    DynamicArray a(2);
    a.push_back(1);
    a.push_back(2);
    a.push_back(3);
    assert(a.size() == 3 && a.at(2) == 3);
    a.insert(1, 99);
    assert(a.at(1) == 99);
    a.erase(1);
    assert(a.at(1) == 2);
    assert(a.pop_back() == 3);
    assert(a.size() == 2);
    cout << "DynamicArray OK" << endl;
    return 0;
}

File: python/data_structures/linear/array/dynamic_array.py
"""手写动态数组（ArrayList 思想）：显式容量与扩容。"""

from __future__ import annotations


class DynamicArray:
    def __init__(self, initial_capacity: int = 4) -> None:
        if initial_capacity < 1:
            initial_capacity = 1
        self._capacity = initial_capacity
        self._size = 0
        self._data: list[object | None] = [None] * self._capacity

    def __len__(self) -> int:
        return self._size

    def capacity(self) -> int:
        return self._capacity

    def at(self, index: int) -> object:
        if index < 0 or index >= self._size:
            raise IndexError("index out of range")
        return self._data[index]  # type: ignore[return-value]

    def push_back(self, value: object) -> None:
        if self._size == self._capacity:
            self._resize(self._capacity * 2)
        self._data[self._size] = value
        self._size += 1

    def pop_back(self) -> object:
        if self._size == 0:
            raise IndexError("pop from empty array")
        self._size -= 1
        v = self._data[self._size]
        self._data[self._size] = None
        return v  # type: ignore[return-value]

    def insert(self, index: int, value: object) -> None:
        if index < 0 or index > self._size:
            raise IndexError("insert index out of range")
        if self._size == self._capacity:
            self._resize(self._capacity * 2)
        for i in range(self._size, index, -1):
            self._data[i] = self._data[i - 1]
        self._data[index] = value
        self._size += 1

    def erase(self, index: int) -> None:
        if index < 0 or index >= self._size:
            raise IndexError("erase index out of range")
        for i in range(index, self._size - 1):
            self._data[i] = self._data[i + 1]
        self._size -= 1
        self._data[self._size] = None

    def _resize(self, new_cap: int) -> None:
        new_data: list[object | None] = [None] * new_cap
        for i in range(self._size):
            new_data[i] = self._data[i]
        self._data = new_data
        self._capacity = new_cap


if __name__ == "__main__":
    a = DynamicArray(2)
    a.push_back(1)
    a.push_back(2)
    a.push_back(3)
    assert len(a) == 3 and a.at(2) == 3
    a.insert(1, 99)
    assert a.at(1) == 99
    a.erase(1)
    assert a.at(1) == 2
    assert a.pop_back() == 3
    assert len(a) == 2
    empty = DynamicArray()
    try:
        empty.pop_back()
        raise AssertionError("expected IndexError")
    except IndexError:
        pass
    try:
        empty.at(0)
        raise AssertionError("expected IndexError")
    except IndexError:
        pass
    print("DynamicArray OK", list(a.at(i) for i in range(len(a))))
